File: src/adapters/driver/http/handlers/frontendHandler_test.go
package handlers

import (
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"
)

func TestFrontendHandler_CacheHeaders(t *testing.T) {
	// Create a temp directory with test files
	dir := t.TempDir()
	files := map[string]string{
		"index.html": "<html></html>",
		"app.js":     "console.log('hello')",
		"style.css":  "body {}",
		"logo.png":   "PNG",
		"favicon.ico": "ICO",
		"data.json":  "{}",
	}
	for name, content := range files {
		if err := os.WriteFile(filepath.Join(dir, name), []byte(content), 0644); err != nil {
			t.Fatalf("failed to create test file %s: %v", name, err)
		}
	}

	handler := FrontendHandler(dir)

	tests := []struct {
		name           string
		path           string
		expectedCache  string
	}{
		{"html", "/index.html", "public, max-age=600"},
		{"js", "/app.js", "public, max-age=86400"},
		{"css", "/style.css", "public, max-age=86400"},
		{"png", "/logo.png", "public, max-age=31536000"},
		{"ico", "/favicon.ico", "public, max-age=31536000"},
		{"unknown ext", "/data.json", "no-cache, no-store, must-revalidate"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			req := httptest.NewRequest("GET", tt.path, nil)
			w := httptest.NewRecorder()

			handler.ServeHTTP(w, req)

			cacheControl := w.Header().Get("Cache-Control")
			if cacheControl != tt.expectedCache {
				t.Errorf("Cache-Control for %s = %q, want %q", tt.path, cacheControl, tt.expectedCache)
			}
		})
	}
}

func TestFrontendHandler_NoCacheExtraHeaders(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "data.json"), []byte("{}"), 0644); err != nil {
		t.Fatal(err)
	}

	handler := FrontendHandler(dir)
	req := httptest.NewRequest("GET", "/data.json", nil)
	w := httptest.NewRecorder()

	handler.ServeHTTP(w, req)

	if w.Header().Get("Pragma") != "no-cache" {
		t.Errorf("expected Pragma: no-cache, got %q", w.Header().Get("Pragma"))
	}
	if w.Header().Get("Expires") != "0" {
		t.Errorf("expected Expires: 0, got %q", w.Header().Get("Expires"))
	}
}

func TestFrontendHandler_ServesFiles(t *testing.T) {
	dir := t.TempDir()
	content := "console.log('hello')"
	if err := os.WriteFile(filepath.Join(dir, "app.js"), []byte(content), 0644); err != nil {
		t.Fatal(err)
	}

	handler := FrontendHandler(dir)
	req := httptest.NewRequest("GET", "/app.js", nil)
	w := httptest.NewRecorder()

	handler.ServeHTTP(w, req)

	if w.Code != http.StatusOK {
		t.Errorf("expected status 200, got %d", w.Code)
	}
}

File: src/test/java/com/fjourdren/theatrum/infrastructure/adapter/in/web/handlers/FrontendHandlerTest.java
package com.fjourdren.theatrum.infrastructure.adapter.in.web.handlers;

import com.fjourdren.theatrum.domain.model.AppPaths;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.springframework.mock.web.MockHttpServletResponse;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import static org.assertj.core.api.Assertions.assertThat;

/** Port of Go's {@code frontendHandler_test.go}. */
class FrontendHandlerTest {

    @TempDir
    Path tempDir;

    private FrontendHandler handler;
    private MockHttpServletResponse response;

    @BeforeEach
    void setUp() throws IOException {
        Path frontendDir = tempDir.resolve("frontend");
        Files.createDirectories(frontendDir);
        Files.writeString(frontendDir.resolve("index.html"), "<html></html>");
        Files.writeString(frontendDir.resolve("app.js"), "console.log('hello')");
        Files.writeString(frontendDir.resolve("style.css"), "body {}");
        Files.writeString(frontendDir.resolve("logo.png"), "PNG");
        Files.writeString(frontendDir.resolve("favicon.ico"), "ICO");
        Files.writeString(frontendDir.resolve("data.json"), "{}");

        handler = new FrontendHandler(new AppPaths(tempDir.resolve("data"), frontendDir));
        response = new MockHttpServletResponse();
    }

    @ParameterizedTest(name = "{0}")
    @CsvSource({
            "/index.html,   'public, max-age=600'",
            "/app.js,       'public, max-age=86400'",
            "/style.css,    'public, max-age=86400'",
            "/logo.png,     'public, max-age=31536000'",
            "/favicon.ico,  'public, max-age=31536000'",
            "/data.json,    'no-cache, no-store, must-revalidate'",
    })
    void cacheHeaders(String path, String expectedCacheControl) throws IOException {
        handler.handle(path, response);

        assertThat(response.getHeader("Cache-Control")).isEqualTo(expectedCacheControl);
    }

    @Test
    void noCacheResponsesCarryPragmaAndExpires() throws IOException {
        handler.handle("/data.json", response);

        assertThat(response.getHeader("Pragma")).isEqualTo("no-cache");
        assertThat(response.getHeader("Expires")).isEqualTo("0");
    }

    @Test
    void servesFiles() throws IOException {
        handler.handle("/app.js", response);

        assertThat(response.getStatus()).isEqualTo(200);
        assertThat(response.getContentAsString()).isEqualTo("console.log('hello')");
    }

    @Test
    void servesIndexHtmlForTheRootPath() throws IOException {
        handler.handle("/", response);

        assertThat(response.getStatus()).isEqualTo(200);
        assertThat(response.getContentAsString()).isEqualTo("<html></html>");
    }

    @Test
    void missingFileReturns404() throws IOException {
        handler.handle("/nope.js", response);

        assertThat(response.getStatus()).isEqualTo(404);
    }

    @Test
    void pathTraversalOutsideTheFrontendDirectoryReturns404() throws IOException {
        Files.writeString(tempDir.resolve("secret.txt"), "top secret");

        handler.handle("/../secret.txt", response);

        assertThat(response.getStatus()).isEqualTo(404);
        assertThat(response.getContentAsString()).doesNotContain("top secret");
    }
}
